File: BRAVO/BRAVORequestAPI.py

import requests
import json
import os
import pickle as pkl
import datetime

DefaultConfigurations = {
    "TimeSeriesRecording": {
        "StandardFilter": {
            "name": "Standard Bandpass Filter",
            "description": "",
            "options": ["No Filter","Butterworth 1-100Hz"],
            "value": "No Filter"
        },
        "NotchFilter": {
            "name": "Powerline Noise Notch Filter",
            "description": "",
            "options": ["No Filter","Notch 55-65Hz","Notch 45-55Hz"],
            "value": "No Filter"
        },
        "WienerFilter": {
            "name": "Wiener Filter for Artifact Removals",
            "description": "",
            "options": ["No Filter","Use Wiener Filter"],
            "value": "No Filter"
        },
        "CardiacFilter": {
            "name": "Cardiac Filter for EKG Removals",
            "description": "",
            "options": ["No Filter","Use Adaptive Template Matching"],
            "value": "No Filter"
        },
        "SpectrogramMethod": {
            "name": "Time-Frequency Analysis Algorithm",
            "description": "",
            "options": ["Welch's Periodogram","Short-time Fourier Transform","Wavelet","Autoregressive Model (Yule-Walker)"],
            "value": "Welch's Periodogram"
        },
        "BaselineCorrection": {
            "name": "Baseline Correlation for Time-Frequency Analysis",
            "description": "",
            "options": ["No Correction"],
            "value": "No Correction"
        },
        "Normalization": {
            "name": "Normalization for Time-Frequency Analysis",
            "description": "",
            "options": ["No Normalization", "1/f PSD Trend Removal"],
            "value": "No Normalization"
        },
    },
    "PowerSpectralDensity": {
        "PSDMethod": {
            "name": "Power Spectrum Estimation Algorithm",
            "description": "",
            "options": ["Estimated Medtronic PSD","Welch's Periodogram","Autoregressive Model (Yule-Walker)","Short-time Fourier Transform"],
            "value": "Welch's Periodogram"
        },
        "MonopolarEstimation": {
            "name": "Monopolar Estimation Algorithm",
            "description": "",
            "options": ["No Estimation", "DETEC Algorithm (Strelow et. al., 2022)"],
            "value": "No Estimation"
        },
    }
}

class BRAVOPlatformRequest:
    def __init__(self, api_key, server="http://localhost"):
        self.__Server = server
        self.__request = requests.Session()
        self.__API_Key = api_key

    def query(self, url, data=None, files=None, content_type="application/json"):
        if not content_type:
            Headers = {"X-Secure-API-Key": self.__API_Key}
        else:
            Headers = {"Content-Type": content_type, "X-Secure-API-Key": self.__API_Key}

        if data:
            return self.__request.post(self.__Server + url,
                                       json.dumps(data) if content_type else data,
                                       headers=Headers)
        elif files:
            return self.__request.post(self.__Server + url,
                                       files=files,
                                       headers=Headers)
        else:
            return self.__request.post(self.__Server + url,
                                       headers=Headers)
    
    def GetUserInfo(self):
        response = self.query("/api/queryProfile")
        if response.status_code == 200:
            payload = response.json()
            self.User = payload
            return payload
        else:
            if response.status_code == 400:
                raise Exception(f"Network Error: {response.json()}")
            else:
                raise Exception(f"Network Error: {response.status_code}")
    
    def QueryParticipants(self):
        response = self.query("/api/queryParticipants")
        if response.status_code == 200:
            payload = response.json()
            return payload
        else:
            if response.status_code == 400:
                raise Exception(f"Network Error: {response.json()}")
            else:
                raise Exception(f"Network Error: {response.status_code}")
    
    def QueryParticipantInformation(self, participant_uid):
        form = {"ParticipantId": participant_uid}
        response = self.query("/api/queryParticipants", data=form)
        if response.status_code == 200:
            payload = response.json()
            return payload
        else:
            if response.status_code == 400:
                raise Exception(f"Network Error: {response.json()}")
            else:
                raise Exception(f"Network Error: {response.status_code}")
    
    def CreateNewStudy(self, study_name):
        form = {"RequestType": "CreateStudy", "StudyName": study_name}
        response = self.query("/api/manageStudyInformation", data=form)
        if response.status_code == 200:
            payload = response.json()
            return payload
        else:
            if response.status_code == 400:
                raise Exception(f"Network Error: {response.json()}")
            else:
                raise Exception(f"Network Error: {response.status_code}")
    
    def AddToStudy(self, study_id, participant_uid):
        form = {"RequestType": "AddParticipant", "StudyId": study_id, "ParticipantId": participant_uid}
        response = self.query("/api/manageStudyInformation", data=form)
        if response.status_code == 200:
            return True
        else:
            if response.status_code == 400:
                raise Exception(f"Network Error: {response.json()}")
            else:
                raise Exception(f"Network Error: {response.status_code}")
    
    def QueryTherapyHistory(self, participant_uid):
        form = {"ParticipantId": participant_uid}
        response = self.query("/api/queryTherapyHistory", data=form)
        if response.status_code == 200:
            payload = response.json()
            return payload
        else:
            if response.status_code == 400:
                raise Exception(f"Network Error: {response.json()}")
            else:
                raise Exception(f"Network Error: {response.status_code}")

    def QueryTherapeuticEffectAnalysis(self, participant_uid, analysis_uid=None, config=None):
        form = {"ParticipantId": participant_uid, "RequestType": "Overview"}
        if analysis_uid:
            form["RequestType"] = "RequestData"
            form["AnalysisId"] = analysis_uid
            form["ActiveChannels"] = "RequestAllChannel"
            
        if config:
            form["ProcessingConfiguration"] = config
            
        response = self.query("/api/queryTherapeuticEffectAnalysis", data=form)
        if response.status_code == 200:
            payload = response.json()
            return payload
        else:
            if response.status_code == 400:
                raise Exception(f"Network Error: {response.json()}")
            else:
                raise Exception(f"Network Error: {response.status_code}")
    
    def QueryTimeSeriesAnalysis(self, participant_uid, recording_uid=None, config=None ):
        form = {"ParticipantId": participant_uid, "RequestType": "Overview"}
        if recording_uid:
            form["RequestType"] = "RequestData"
            form["AnalysisId"] = recording_uid
            form["ActiveChannels"] = "RequestAllChannel"

        if config:
            form["ProcessingConfiguration"] = config
            
        response = self.query("/api/queryTimeseriesAnalysis", data=form)
        if response.status_code == 200:
            payload = response.json()
            return payload
        else:
            if response.status_code == 400:
                raise Exception(f"Network Error: {response.json()}")
            else:
                raise Exception(f"Network Error: {response.status_code}")
    
    def QueryChronicNeuralActivity(self, participant_uid):
        form = {"ParticipantId": participant_uid, "RequestType": "RequestAll"}
        response = self.query("/api/queryChronicNeuralActivity", data=form)
        if response.status_code == 200:
            payload = response.json()
            return payload
        else:
            if response.status_code == 400:
                raise Exception(f"Network Error: {response.json()}")
            else:
                raise Exception(f"Network Error: {response.status_code}")
    
    def QueryChronicTimeline(self, participant_uid):
        form = {"ParticipantId": participant_uid, "RequestType": "RequestAll"}
        response = self.query("/api/queryChronicTimeline", data=form)
        if response.status_code == 200:
            payload = response.json()
            return payload
        else:
            if response.status_code == 400:
                raise Exception(f"Network Error: {response.json()}")
            else:
                raise Exception(f"Network Error: {response.status_code}")
    
    def QueryRawTimeseries(self, participant_uid, recording_uid=None ):
        form = {"ParticipantId": participant_uid, "RequestType": "Overview"}
        if recording_uid:
            form["RequestType"] = "RawTimeseries"
            form["RecordingId"] = recording_uid
            
        response = self.query("/api/queryRawTimeseries", data=form)
        if response.status_code == 200:
            payload = response.json()
            return payload
        else:
            if response.status_code == 400:
                raise Exception(f"Network Error: {response.json()}")
            else:
                raise Exception(f"Network Error: {response.status_code}")
    
    def UploadMedtronicJSON(self, participant, file, metadata={"device_location": "", "infer_from_device": True}):
        form = {
            "File": file,
            "DataType": (None, "MedtronicJSON"), 
            "ParticipantId": (None, participant), 
            "Institute": (None, self.User["Institute"]), 
            "Metadata": (None, json.dumps(metadata))
        }

        response = self.query("/api/uploadData", files=form, content_type=None)
        if response.status_code == 200:
            return True
        elif response.status_code == 301:
            return True
        else:
            if response.status_code == 400:
                raise Exception(f"Network Error: {response.json()}")
            else:
                raise Exception(f"Network Error: {response.status_code}")
            

File: BRAVO/BRAVORequestAPI.m
classdef BRAVORequestAPI
    % Request APIs to interface with BRAVO Platform (v3.0.0-alpha)
    %   Detailed explanation goes here
    
    properties
        APIKey
        ServerAddress
        Profile
    end
    
    methods
        function requester = BRAVORequestAPI(key,server)
            arguments
                key char
                server char = 'http://localhost:27286'
            end
            
            requester.APIKey = key;
            requester.ServerAddress = server;
            requester.Profile = GetUserInfo(requester);
        end
        
        %{  
        Example Usage:
            Profile = requester.GetUserInfo();
        %}
        function response = GetUserInfo(requester)
            options = weboptions('HeaderFields', {'X-Secure-API-Key' requester.APIKey;}, 'Timeout', 5);
            response = webwrite([requester.ServerAddress '/api/queryProfile'], options);
        end
        
        %{  
        Example Usage:
            Config = requester.QueryAnalysisConfiguration();
        %}
        function response = QueryAnalysisConfiguration(requester, params)
            arguments
                requester
                params.config = false
            end

            data.Configurations = struct();
            if isa(params.config, 'struct')
                data.Configurations = params.config;
            end

            options = weboptions('HeaderFields', {'X-Secure-API-Key' requester.APIKey;}, 'Timeout', 5);
            response = webwrite([requester.ServerAddress '/api/queryProfile'], data, options);
        end
        
        function response = DeleteParticipant(requester, participant_uid)
            arguments
                requester
                participant_uid
            end

            data.ParticipantId = participant_uid;

            options = weboptions('HeaderFields', {'X-Secure-API-Key' requester.APIKey;}, 'Timeout', 500);
            response = webwrite([requester.ServerAddress '/api/deleteParticipantInformation'], data, options);
        end
        
        function response = AddNewParticipant(requester, participant_name)
            arguments
                requester
                participant_name
            end

            data.Name = participant_name;

            options = weboptions('HeaderFields', {'X-Secure-API-Key' requester.APIKey;}, 'Timeout', 5);
            response = webwrite([requester.ServerAddress '/api/createParticipantInformation'], data, options);
        end
        
        function response = AddNewDBSDevice(requester, participant_uid, device_type, device_serial)
            arguments
                requester
                participant_uid
                device_type
                device_serial
            end

            data.ParticipantId = participant_uid;
            data.DeviceType = "DBSDevice";
            data.ManufacturerDeviceType = device_type;
            data.SerialNumber = device_serial;

            options = weboptions('HeaderFields', {'X-Secure-API-Key' requester.APIKey;}, 'Timeout', 5);
            response = webwrite([requester.ServerAddress '/api/manageParticipantDevice'], data, options);
        end
        
        function response = QueryParticipants(requester)
            options = weboptions('HeaderFields', {'X-Secure-API-Key' requester.APIKey;}, 'Timeout', 5);
            response = webwrite([requester.ServerAddress '/api/queryParticipants'], options);
        end
        
        function response = QueryParticipantInformation(requester, participant_uid)
            arguments
                requester
                participant_uid
            end

            data.ParticipantId = participant_uid;

            options = weboptions('HeaderFields', {'X-Secure-API-Key' requester.APIKey;}, 'Timeout', 5);
            response = webwrite([requester.ServerAddress '/api/queryParticipantInformation'], data, options);
        end
        
        % This include chronic events and patient controller events
        function response = QueryParticipantEvents(requester, participant_uid)
            arguments
                requester
                participant_uid
            end

            data.ParticipantId = participant_uid;

            options = weboptions('HeaderFields', {'X-Secure-API-Key' requester.APIKey;}, 'Timeout', 5);
            response = webwrite([requester.ServerAddress '/api/queryParticipantEvents'], data, options);
        end
        
        % This include chronic annotations and streaming annotations
        function response = QueryParticipantAnnotations(requester, participant_uid)
            arguments
                requester
                participant_uid
            end

            data.ParticipantId = participant_uid;

            options = weboptions('HeaderFields', {'X-Secure-API-Key' requester.APIKey;}, 'Timeout', 5);
            response = webwrite([requester.ServerAddress '/api/queryParticipantAnnotations'], data, options);
        end
        
        function response = QueryParticipantSurveyRecords(requester, participant_uid, form_uid)
            arguments
                requester
                participant_uid
                form_uid = false
            end

            data.ParticipantId = participant_uid;
            data.RequestType = 'RequestAll';
            
            if isa(form_uid, 'char')
                data.RequestType = 'RequestRecords';
                data.FormId = form_uid;
            end

            options = weboptions('HeaderFields', {'X-Secure-API-Key' requester.APIKey;}, 'Timeout', 5);
            response = webwrite([requester.ServerAddress '/api/queryParticipantSurveyRecords'], data, options);
        end
        
        function response = QueryTherapeuticEffectAnalysis(requester, participant_uid, params)
            arguments
                requester
                participant_uid
                params.analysis_uid = false
                params.config = false
                params.refresh = false
            end

            data.ParticipantId = participant_uid;
            data.RequestType = 'Overview';

            if params.refresh
                data.RequestType = "DeleteCache";
                data.AnalysisId = params.analysis_uid;
                options = weboptions('HeaderFields', {'X-Secure-API-Key' requester.APIKey;}, 'Timeout', 60);
                webwrite([requester.ServerAddress '/api/queryTherapeuticEffectAnalysis'], data, options);

                params.refresh = false;
                params = namedargs2cell(params);
                response = QueryTherapeuticEffectAnalysis(requester, participant_uid, params{:});
                return;
            end

            if isa(params.analysis_uid, 'char') || isa(params.analysis_uid, 'string')
                data.RequestType = 'RequestData';
                data.AnalysisId = params.analysis_uid;
                data.ActiveChannels = 'RequestAllChannel';
            end

            if isa(params.config, 'struct')
                data.ProcessingConfiguration = params.config;
            end

            options = weboptions('HeaderFields', {'X-Secure-API-Key' requester.APIKey;}, 'Timeout', 60);
            response = webwrite([requester.ServerAddress '/api/queryTherapeuticEffectAnalysis'], data, options);
        end

        function response = QueryTimeseriesAnalysis(requester, participant_uid, params)
            arguments
                requester
                participant_uid
                params.analysis_uid = false
                params.therapy_uid = false
                params.config = false
                params.refresh = false
            end

            data.ParticipantId = participant_uid;
            data.RequestType = 'Overview';
            data.TherapyId = params.therapy_uid;

            if params.refresh
                data.RequestType = "DeleteCache";
                data.AnalysisId = params.analysis_uid;
                data.TherapyId = params.therapy_uid;
                options = weboptions('HeaderFields', {'X-Secure-API-Key' requester.APIKey;}, 'Timeout', 60);
                webwrite([requester.ServerAddress '/api/queryTimeseriesAnalysis'], data, options);

                params.refresh = false;
                params = namedargs2cell(params);
                response = QueryTimeseriesAnalysis(requester, participant_uid, params{:});
                return;
            end

            if isa(params.analysis_uid, 'char') || isa(params.analysis_uid, 'string')
                data.RequestType = 'RequestData';
                data.AnalysisId = params.analysis_uid;
                data.ActiveChannels = 'RequestAllChannel';
            end

            if isa(params.config, 'struct')
                data.ProcessingConfiguration = params.config;
            end

            options = weboptions('HeaderFields', {'X-Secure-API-Key' requester.APIKey;}, 'Timeout', 60);
            response = webwrite([requester.ServerAddress '/api/queryTimeseriesAnalysis'], data, options);
        end
        
        function response = QueryChronicNeuralActivity(requester, participant_uid, params)
            arguments
                requester
                participant_uid
                params.refresh = false
            end

            data.ParticipantId = participant_uid;
            data.RequestType = 'RequestAll';

            if params.refresh
                data.RequestType = "DeleteCache";
                options = weboptions('HeaderFields', {'X-Secure-API-Key' requester.APIKey;}, 'Timeout', 60);
                webwrite([requester.ServerAddress '/api/queryChronicNeuralActivity'], data, options);

                params.refresh = false;
                params = namedargs2cell(params);
                response = QueryChronicNeuralActivity(requester, participant_uid, params{:});
                return;
            end

            options = weboptions('HeaderFields', {'X-Secure-API-Key' requester.APIKey;}, 'Timeout', 60);
            response = webwrite([requester.ServerAddress '/api/queryChronicNeuralActivity'], data, options);
        end
        
        function response = QueryChronicTimeline(requester, participant_uid, params)
            arguments
                requester
                participant_uid
                params.refresh = false
            end

            data.ParticipantId = participant_uid;
            data.RequestType = 'RequestAll';

            if params.refresh
                data.RequestType = "DeleteCache";
                options = weboptions('HeaderFields', {'X-Secure-API-Key' requester.APIKey;}, 'Timeout', 60);
                webwrite([requester.ServerAddress '/api/queryChronicTimeline'], data, options);

                params.refresh = false;
                params = namedargs2cell(params);
                response = QueryChronicTimeline(requester, participant_uid, params{:});
                return;
            end

            options = weboptions('HeaderFields', {'X-Secure-API-Key' requester.APIKey;}, 'Timeout', 60);
            response = webwrite([requester.ServerAddress '/api/queryChronicTimeline'], data, options);
        end
        
        function response = QueryNeuralActivitySnapshot(requester, participant_uid, params)
            arguments
                requester
                participant_uid
                params.refresh = false
                params.config = false
            end

            data.ParticipantId = participant_uid;
            data.RequestType = 'RequestAll';
            
            if class(params.config) == "struct"
                data.ProcessingConfiguration = params.config;
            end

            if params.refresh
                data.RequestType = "DeleteCache";
                options = weboptions('HeaderFields', {'X-Secure-API-Key' requester.APIKey;}, 'Timeout', 60);
                webwrite([requester.ServerAddress '/api/queryNeuralActivitySnapshot'], data, options);

                params.refresh = false;
                params = namedargs2cell(params);
                response = QueryNeuralActivitySnapshot(requester, participant_uid, params{:});
                return;
            end

            options = weboptions('HeaderFields', {'X-Secure-API-Key' requester.APIKey;}, 'Timeout', 60);
            response = webwrite([requester.ServerAddress '/api/queryNeuralActivitySnapshot'], data, options);
        end
        
        function response = QueryBurstAnalysis(requester, participant_uid, recording_ids, channel, fc, params)
            arguments
                requester
                participant_uid
                recording_ids
                channel
                fc
                params.config = false
            end

            data.ParticipantId = participant_uid;
            data.RequestType = 'RequestData';
            data.RecordingIds = recording_ids;
            data.Channel = channel;
            data.CenterFrequency = fc;

            disp(data);

            if isa(params.config, 'struct')
                data.ProcessingConfiguration = params.config;
            end

            options = weboptions('HeaderFields', {'X-Secure-API-Key' requester.APIKey;}, 'Timeout', 60);
            response = webwrite([requester.ServerAddress '/api/queryBurstAnalysis'], data, options);
        end
        
        function response = QueryTherapyHistory(requester, participant_uid)
            arguments
                requester
                participant_uid
            end

            data.ParticipantId = participant_uid;
            options = weboptions('HeaderFields', {'X-Secure-API-Key' requester.APIKey;}, 'Timeout', 60);
            response = webwrite([requester.ServerAddress '/api/queryTherapyHistory'], data, options);
        end

        function response = QueryMedicationCycleAnalysis(requester, participant_uid, params)
            arguments
                requester
                participant_uid
                params.recording_ids = false
            end

            data.ParticipantId = participant_uid;
            data.RequestType = 'RequestAll';

            if isa(params.recording_ids, 'cell')
                data.RequestType = 'RequestAnalysis';
                data.RecordingIds = params.recording_ids;
            end

            options = weboptions('HeaderFields', {'X-Secure-API-Key' requester.APIKey;}, 'Timeout', 60);
            response = webwrite([requester.ServerAddress '/api/queryMedicationCycleAnalysis'], data, options);
        end

        function response = RequestAIPrediction(requester, analysis_name, input)
            arguments
                requester
                analysis_name,
                input
            end
            
            data.RequestType = 'RequestAnalysis';
            data.AnalysisName = analysis_name;
            data.Data = input;

            options = weboptions('HeaderFields', {'X-Secure-API-Key' requester.APIKey;}, 'Timeout', 60);
            response = webwrite([requester.ServerAddress '/api/requestAIPrediction'], data, options);
        end
    end
end

